### PR TITLE
feat: make momentum-ness infectious

### PR DESCRIPTION
**Pull request by Saransh-cpp**
## Description

**Kindly take a look at [CONTRIBUTING.md](https://github.com/scikit-hep/vector/blob/main/.github/CONTRIBUTING.md).**

_Please describe the purpose of this pull request. Reference and link to any relevant issues or pull requests._

## Checklist

- [ ] Have you followed the guidelines in our Contributing document?
- [ ] Have you checked to ensure there aren't any other open Pull Requests for the required change?
- [ ] Does your submission pass pre-commit? (`$ pre-commit run --all-files` or `$ nox -s lint`)
- [ ] Does your submission pass tests? (`$ pytest` or `$ nox -s tests`)
- [ ] Does the documentation build with your changes? (`$ cd docs; make clean; make html` or `$ nox -s docs`)
- [ ] Does your submission pass the doctests? (`$ pytest --doctest-plus src/vector/` or `$ nox -s doctests`)

## Before Merging

- [ ] Summarize the commit messages into a brief review of the Pull request.

### CODE REVIEW

File: tests/compute/test_conversions.py [hunk 1]
```diff
@@ -862,30 +862,72 @@ def test_like_numpy():
...
     # 2D + 4D.like(2D) = 2D
-    assert all(v1 + v3.like(v1) == pv1_v2)
-    assert all(v3.like(v1) + v1 == pv1_v2)
+    assert isinstance(v1 + v3.like(v1), vector.MomentumNumpy2D)
+    assert isinstance(v3.like(v1) + v1, vector.MomentumNumpy2D)
     # 3D + 2D.like(3D) = 3D
-    assert all(v2 + v1.like(v2) == pv2_v1)
-    assert all(v1.like(v2) + v2 == pv2_v1)
+    assert isinstance(v2 + v1.like(v2), vector.backends.numpy.MomentumNumpy3D)
+    assert isinstance(v1.like(v2) + v2, vector.backends.numpy.MomentumNumpy3D)
```
Comment (Saransh-cpp) [inline tests/compute/test_conversions.py:924]:
```suggestion
    assert isinstance(v2 + v1.like(v2), vector.MomentumNumpy3D)
    assert isinstance(v1.like(v2) + v2, vector.MomentumNumpy3D)
```

File: tests/compute/test_conversions.py [hunk 2]
```diff
@@ -862,30 +862,72 @@ def test_like_numpy():
...
     # 3D + 4D.like(3D) = 3D
-    assert all(v2 + v3.like(v2) == pv2_v3)
-    assert all(v3.like(v2) + v2 == pv2_v3)
+    assert isinstance(v2 + v3.like(v2), vector.backends.numpy.MomentumNumpy3D)
+    assert isinstance(v3.like(v2) + v2, vector.backends.numpy.MomentumNumpy3D)
```
Comment (Saransh-cpp) [inline tests/compute/test_conversions.py:927]:
```suggestion
    assert isinstance(v2 + v3.like(v2), vector.MomentumNumpy3D)
    assert isinstance(v3.like(v2) + v2, vector.MomentumNumpy3D)
```

File: tests/compute/test_conversions.py [hunk 3]
```diff
@@ -862,30 +862,72 @@ def test_like_numpy():
...
     # 4D + 2D.like(4D) = 4D
-    assert all(v3 + v1.like(v3) == pv1_v3)
-    assert all(v1.like(v3) + v3 == pv1_v3)
+    assert isinstance(v3 + v1.like(v3), vector.backends.numpy.MomentumNumpy4D)
+    assert isinstance(v1.like(v3) + v3, vector.backends.numpy.MomentumNumpy4D)
```
Comment (Saransh-cpp) [inline tests/compute/test_conversions.py:930]:
```suggestion
    assert isinstance(v3 + v1.like(v3), vector.MomentumNumpy4D)
    assert isinstance(v1.like(v3) + v3, vector.MomentumNumpy4D)
```

File: tests/compute/test_conversions.py [hunk 4]
```diff
@@ -862,30 +862,72 @@ def test_like_numpy():
...
     # 4D + 3D.like(4D) = 4D
-    assert all(v3 + v2.like(v3) == pv3_v2)
-    assert all(v2.like(v3) + v3 == pv3_v2)
+    assert isinstance(v3 + v2.like(v3), vector.backends.numpy.MomentumNumpy4D)
+    assert isinstance(v2.like(v3) + v3, vector.backends.numpy.MomentumNumpy4D)
```
Comment (Saransh-cpp) [inline tests/compute/test_conversions.py:933]:
```suggestion
    assert isinstance(v3 + v2.like(v3), vector.MomentumNumpy4D)
    assert isinstance(v2.like(v3) + v3, vector.MomentumNumpy4D)
```

File: src/vector/_methods.py [hunk 5]
```diff
@@ -4405,6 +4362,6 @@ def _flavor_of(*objects: VectorProtocol) -> type[VectorProtocol]:
 
     assert handler is not None
     if is_momentum:
-        return type(handler)
+        return handler.MomentumClass
```
Comment (Saransh-cpp) [inline src/vector/_methods.py:4365]: I tried doing this without creating `MomentumClass`, but that introduced a lot of edge cases and if conditions in the `_flavor_of` function. Using `MomentumClass` makes it clean and uniform for all the backends.

**Review by jpivarski — APPROVED** [review]
This looks good to me! I think you can merge this.

**Pull request merged**